File: native/android/core/src/main/java/com/vocare/translate/core/api/VocaApi.kt
package com.vocare.translate.core.api

import com.vocare.translate.core.model.Balance
import com.vocare.translate.core.model.CreateSessionRequest
import com.vocare.translate.core.model.IceServer
import com.vocare.translate.core.model.OfferAnswer
import com.vocare.translate.core.model.OfferRequest
import com.vocare.translate.core.model.PollResponse
import com.vocare.translate.core.model.PttAction
import com.vocare.translate.core.model.PttResult

/**
 * Typed failures from the backend. Every [VocaApi] method that can throw throws
 * one of these (never a raw IOException), so callers route on type.
 */
sealed class ApiException(message: String, cause: Throwable? = null) : Exception(message, cause) {
    /** 402 from `/offer`: metering is enforced and the balance is 0. */
    class PaymentRequired(val balance: Balance?) : ApiException("no_translation_credit")

    /** Any other non-2xx. [error] is the backend's `{ "error": ... }` when it sent one. */
    class Http(val code: Int, val error: String?) : ApiException("http_$code" + (error?.let { ": $it" } ?: ""))

    /** Transport failure: DNS, timeout, connection reset, TLS. */
    class Network(cause: Throwable) : ApiException("network_error", cause)

    /** 2xx with a body we could not parse. */
    class Decode(cause: Throwable) : ApiException("decode_error", cause)

    /** `/api/ice` returned an empty list — the session must fail before any offer (CONTRACT §2). */
    class NoIceServers : ApiException("no_ice_servers")
}

/**
 * The backend, as `:session` sees it (CONTRACT §6). Implemented in `:app`
 * (Retrofit) and injected. Entitlement/history endpoints the session does not
 * need live on `:app`'s own extension of this interface.
 */
interface VocaApi {
    /** @throws ApiException.NoIceServers when the list is empty. */
    suspend fun iceServers(): List<IceServer>

    /** @return the new `session_id`. */
    suspend fun createSession(req: CreateSessionRequest): String

    /** @throws ApiException.PaymentRequired on 402. */
    suspend fun offer(req: OfferRequest): OfferAnswer

    suspend fun ptt(pcId: String, action: PttAction): PttResult

    suspend fun poll(sessionId: String): PollResponse

    /** Fire-and-forget; never throws. */
    suspend fun hangup(pcId: String)

    /** Reports the session's *total* elapsed seconds. Never throws; null when the report failed. */
    suspend fun consume(subject: String, sessionId: String, seconds: Int): Balance?
}

File: native/ios/VocaKit/Sources/VocaKit/API/VocaAPI.swift
import Foundation

// MARK: - CONTRACT §6 — implemented in VocaKit/API, injected into VocaSession.

public protocol VocaAPI: Sendable {
    func iceServers() async throws -> [ICEServer]
    /// Returns the `session_id`.
    func createSession(_ req: CreateSessionRequest) async throws -> String
    /// Throws `APIError.paymentRequired(balance)` on 402.
    func offer(_ req: OfferRequest) async throws -> OfferAnswer
    func ptt(pcId: String, action: PTTAction) async throws -> PTTResult
    func poll(sessionId: String) async throws -> PollResponse
    /// Fire-and-forget; never throws.
    func hangup(pcId: String) async
    /// Reports the session's TOTAL elapsed seconds (idempotent watermark).
    /// Returns nil on any failure — usage reporting never breaks a session.
    func consume(subject: String, sessionId: String, seconds: Int) async -> Balance?
}

/// The endpoints only the app shell (VocaKit) needs: entitlement and history.
/// Kept separate so `VocaSession` is handed exactly the surface in §6.
public protocol VocaEntitlementAPI: Sendable {
    func entitlement(subject: String) async throws -> Balance
    func activate(_ req: ActivateRequest) async throws -> Balance
}

public protocol VocaHistoryAPI: Sendable {
    func sessions(clientId: String) async throws -> [SessionSummary]
    func session(id: String) async throws -> SessionDetail?
}

public typealias VocaFullAPI = VocaAPI & VocaEntitlementAPI & VocaHistoryAPI
